src/JsonReader.cs

```diff
@@ -484,26 +484,64 @@ public static IJsonReader<object> AsObject<T>(this IJsonReader<T> reader)
     public static IJsonReader<T> Either<T>(IJsonReader<T> reader1, IJsonReader<T> reader2) =>
         Either(reader1, reader2, null);
 
+    enum EitherSideState { Incomplete, Error }
+
     public static IJsonReader<T> Either<T>(IJsonReader<T> reader1, IJsonReader<T> reader2,
                                            string? errorMessage) =>
         Create((ref Utf8JsonReader rdr) =>
         {
-            var irdr = rdr;
-            switch (reader1.TryRead(ref irdr))
+            var (leftStack, rightStack, leftState, rightState) = rdr.ResumeOrDefault<(Stack<object>?, Stack<object>?, EitherSideState, EitherSideState)>();
```

Comment:
nit: thoughts on giving this more structure?

```c#
var ((leftStack, leftState), (rightStack, rightState)) =
    rdr.ResumeOrDefault<((Stack<object>?, EitherSideState), (Stack<object>?, EitherSideState))>();
```

or:

```c#
var (left, right) =
    rdr.ResumeOrDefault<((Stack<object>? Stack, EitherSideState State),
                         (Stack<object>? Stack, EitherSideState State))>();
```

The latter can also make the suspension look simpler:

```c#
return rdr.Suspend((left, right));
```

Reply:
I like it, applied your second suggestion with 0f3df136609273047c601e9c54b742692f2efda8